fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubColumnsRule.java

```diff
@@ -0,0 +1,319 @@
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CollectionElementOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+import com.starrocks.sql.optimizer.operator.scalar.SubfieldOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.task.TaskContext;
+import com.starrocks.thrift.TAccessPathType;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class PruneSubColumnsRule implements TreeRewriteRule {
+
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        SubColumnAccessPathComputer computer = new SubColumnAccessPathComputer();
+        return root.getOp().accept(computer, root, null);
+    }
+
+    /*
+     * collect all complex expressions, such as: MAP_KEYS, MAP_VALUES, map['key'], struct.a.b.c ...
+     */
+    private static class ComplexExpressionCollector extends ScalarOperatorVisitor<Void, Void> {
+        private static final List<String> SUPPORT_FUNCTIONS = ImmutableList.<String>builder()
+                .add(FunctionSet.MAP_KEYS, FunctionSet.MAP_SIZE, FunctionSet.MAP_VALUES)
+                .build();
+
+        private final List<ScalarOperator> complexExpressions = Lists.newArrayList();
+
+        @Override
+        public Void visit(ScalarOperator scalarOperator, Void context) {
+            for (ScalarOperator child : scalarOperator.getChildren()) {
+                child.accept(this, context);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitVariableReference(ColumnRefOperator variable, Void context) {
+            complexExpressions.add(variable);
+            return null;
+        }
+
+        @Override
+        public Void visitCollectionElement(CollectionElementOperator collectionElementOp, Void context) {
+            complexExpressions.add(collectionElementOp);
+            return null;
+        }
+
+        @Override
+        public Void visitSubfield(SubfieldOperator subfieldOperator, Void context) {
+            complexExpressions.add(subfieldOperator);
+            return null;
+        }
+
+        @Override
+        public Void visitCall(CallOperator call, Void context) {
+            if (SUPPORT_FUNCTIONS.contains(call.getFnName())) {
+                complexExpressions.add(call);
+                return null;
+            }
+            return visit(call, context);
+        }
+    }
+
+    /*
+     * compute all complex column access path in whole plan, we only need check the expression
+     * which one in project and predicate (the rule execute order promise this)
```

Comment:
What about predicates have been pushed down to operator ?

Reply:
the rule will execute before push down predicate